lib/srv/db/spanner/engine.go

```diff
@@ -0,0 +1,162 @@
+	// where the project ID and instance ID come from the Teleport database's
+	// GCP metadata.
+	// Every RPC that specifies a database name must be prefixed by this or the
+	// RPC will be denied.
+	instanceID string
+
+	// mu guards access to gcloudClient
+	mu sync.Mutex
+	// gcloudConn is an upstream client to Google cloud API used to forward RPCs
+	// to the real Spanner API.
+	gcloudClient gtransport.ConnPool
+
+	// sessionStarted is true after the session start event has been emitted.
+	sessionStarted bool
+	// sessionStartErr is the error (or nil) from the first RPC access check
+	// that starts the session.
+	sessionStartErr error
+	// startSessionOnce is used to start the session only once.
+	startSessionOnce sync.Once
+
+	// UnimplementedSpannerServer is embedded for forward interface compat.
+	spannerpb.UnimplementedSpannerServer
+}
+
+// InitializeConnection initializes the database connection.
+func (e *Engine) InitializeConnection(clientConn net.Conn, sessionCtx *common.Session) error {
+	e.clientConn = clientConn
+	gcp := sessionCtx.Database.GetGCP()
+	e.instanceID = "projects/" + gcp.ProjectID + "/instances/" + gcp.InstanceID + "/"
+	e.sessionCtx = sessionCtx
+	return nil
+}
+
+func (e *Engine) SendError(err error) {
+	if err == nil || utils.IsOKNetworkError(err) {
+		return
+	}
+	// the grpc server handles sending all errors, if an error is sent outside
+	// of that, just log it here.
+	e.Log.WithError(err).Debug("GCP Spanner connection error")
+}
+
+// HandleConnection processes the connection from the proxy coming over reverse
+// tunnel.
+func (e *Engine) HandleConnection(ctx context.Context, _ *common.Session) error {
+	observe := common.GetConnectionSetupTimeObserver(e.sessionCtx.Database)
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	e.clientConnContext = ctx
+	e.cancelContext = cancel
+
+	// stand up a gRPC grpcServer to handle the client connection.
+	grpcServer := grpc.NewServer(
+		grpc.ChainUnaryInterceptor(e.unaryServerInterceptors()...),
+		grpc.ChainStreamInterceptor(e.streamServerInterceptors()...),
+		grpc.MaxConcurrentStreams(defaults.GRPCMaxConcurrentStreams),
+	)
+	// register server services to proxy RPCs.
+	spannerpb.RegisterSpannerServer(grpcServer, e)
+
+	// this doesn't block, because the listener returns when Accept is called
+	// for a second time.
+	err := grpcServer.Serve(newSingleUseListener(e.clientConn, observe))
+	if err != nil && !errors.Is(err, io.EOF) {
+		return trace.Wrap(err)
+	}
+	select {
+	case <-ctx.Done():
+	case <-e.Context.Done():
+	}
+
+	// this will cause the server to reject all new RPCs and block until all
+	// outstanding handlers have returned.
+	grpcServer.GracefulStop()
+	if e.gcloudClient != nil {
+		e.gcloudClient.Close()
+	}
+
+	// emit a session end event if the session was started successfully.
+	if e.sessionStarted && e.sessionStartErr == nil {
+		e.Audit.OnSessionEnd(e.Context, e.sessionCtx)
+	}
+	return nil
```

Comment:
I don't think so - all it will do is end up in the db server's deferred `engine.SendError` which just logs it, but we already logged handler errors in the interceptors

Reply:
Ya. I am not too worried about how errors are propagated back to the client. I am just worried if there is any other logic that relies on the error returned by the `HandleConnection`. For example, the connection test logic seems catching this error. Well, we don't have connection test for spanner but not sure some other logic in the future might peek into these errors.

Reply:
Ok, i'll return any error that causes the grpc server to shutdown early.
That *was* just RBAC errors.
But I think if any error occurs in `e.getClient` it's probably not going to be transient, i.e. it indicates connectivity issue with GCP, or IAM creds issue, so I also made those errors shutdown the server.

Done in: 1253f9a487e44104cb6b2eca7e83dda973bc1428